validation_epoch_end needs to return CUDA tensors

<!-- 
### Common bugs:
1. Tensorboard not showing in Jupyter-notebook see [issue 79](https://github.com/PyTorchLightning/pytorch-lightning/issues/79).    
2. PyTorch 1.1.0 vs 1.2.0 support [see FAQ](https://github.com/PyTorchLightning/pytorch-lightning#faq)    
-->

## 🐛 Bug

<!-- A clear and concise description of what the bug is. -->

I'm not sure if this is expected behaviour or not, but upgrading to the latest version (from 0.8.1) caused my `validation_epoch_end` to break. It appears that a CUDA tensor is expected for the metric where before the tensor was device agnostic.

This was using sklearn's `roc_auc_score`. I haven't yet got around to testing PL's new metrics.

Feel free to close if this is expected behaviour.

### To Reproduce

This is my `validation_epoch_end`. Uncommenting `.to(avg_loss.device)` allows this to run with the dev version of PL.

This was run with `ddp`, `precision=16` using Apex AMP.

```
def validation_epoch_end(self, outputs):
        avg_loss = torch.stack([x["val_loss"] for x in outputs]).mean()
        y_pred = torch.cat([x["y_pred"].squeeze() for x in outputs]).cpu().numpy()
        y_true = torch.cat([x["y_true"].squeeze() for x in outputs]).cpu().numpy()

        metric = torch.tensor(roc_auc_score(y_true, y_pred))  # .to(avg_loss.device)

        tensorboard_logs = {
            "loss/validation": avg_loss,
            "auc": metric,
        }
        return {"val_loss": avg_loss, "log": tensorboard_logs, "auc": metric}
```

<!-- If you have a code sample, error messages, stack traces, please provide it here as well -->
The error message can be seen here: https://github.com/PyTorchLightning/pytorch-lightning/issues/2411#issuecomment-652171874 

#### Code sample
<!-- Ideally attach a minimal code sample to reproduce the decried issue. 
Minimal means having the shortest code but still preserving the bug. -->

### Expected behavior

<!-- A clear and concise description of what you expected to happen. -->

### Environment

Please copy and paste the output from our
[environment collection script](https://raw.githubusercontent.com/PyTorchLightning/pytorch-lightning/master/tests/collect_env_details.py)
(or fill out the checklist below manually).

You can get the script and run it with:
```
wget https://raw.githubusercontent.com/PyTorchLightning/pytorch-lightning/master/tests/collect_env_details.py
# For security purposes, please check the contents of collect_env_details.py before running it.
python collect_env_details.py
```

 - PyTorch Version (e.g., 1.0): 1.5
 - OS (e.g., Linux): Ubuntu 20.04
 - How you installed PyTorch (`conda`, `pip`, source):
 - Build command you used (if compiling from source):
 - Python version: 3.8.2
 - CUDA/cuDNN version: 10.2
 - GPU models and configuration: Dual GPU (ddp)
 - Any other relevant information: Apex AMP

### Additional context

<!-- Add any other context about the problem here. -->

Ok I tried it in trainer, and got this error (after disabling AMP):
```
Loaded pretrained weights for efficientnet-b0
train.py:141: DtypeWarning: Columns (5) have mixed types.Specify dtype option on import or set low_memory=False.
  train_single_fold(args)
GPU available: True, used: False
TPU available: False, using: 0 TPU cores

  | Name      | Type             | Params
-----------------------------------------------
0 | critereon | CrossEntropyLoss | 0     
1 | metric    | AUROC            | 0     
2 | net       | EfficientNet     | 4 M   
Traceback (most recent call last):
  File "train.py", line 141, in <module>
    train_single_fold(args)
  File "train.py", line 65, in train_single_fold
    trainer.fit(model)
  File "/home/anjum/anaconda3/envs/kaggle/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py", line 1020, in fit
    self.run_pretrain_routine(model)
  File "/home/anjum/anaconda3/envs/kaggle/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py", line 1128, in run_pretrain_routine
    self.reset_val_dataloader(ref_model)
  File "/home/anjum/anaconda3/envs/kaggle/lib/python3.8/site-packages/pytorch_lightning/trainer/data_loading.py", line 342, in reset_val_dataloader
    self._reset_eval_dataloader(model, 'val')
  File "/home/anjum/anaconda3/envs/kaggle/lib/python3.8/site-packages/pytorch_lightning/trainer/data_loading.py", line 268, in _reset_eval_dataloader
    dataloaders = self.request_dataloader(getattr(model, f'{mode}_dataloader'))
  File "/home/anjum/anaconda3/envs/kaggle/lib/python3.8/site-packages/pytorch_lightning/trainer/data_loading.py", line 363, in request_dataloader
    dataloader = dataloader_fx()
  File "/home/anjum/PycharmProjects/kaggle/siim_isic_melanoma_classification/engine.py", line 194, in val_dataloader
    sampler = DistributedSampler(dataset)
  File "/home/anjum/anaconda3/envs/kaggle/lib/python3.8/site-packages/torch/utils/data/distributed.py", line 43, in __init__
    num_replicas = dist.get_world_size()
  File "/home/anjum/anaconda3/envs/kaggle/lib/python3.8/site-packages/torch/distributed/distributed_c10d.py", line 582, in get_world_size
    return _get_group_size(group)
  File "/home/anjum/anaconda3/envs/kaggle/lib/python3.8/site-packages/torch/distributed/distributed_c10d.py", line 196, in _get_group_size
    _check_default_pg()
  File "/home/anjum/anaconda3/envs/kaggle/lib/python3.8/site-packages/torch/distributed/distributed_c10d.py", line 186, in _check_default_pg
    assert _default_pg is not None, \
AssertionError: Default process group is not initialized
```

I'm not familiar with gloo so I don't know what I'm looking at here. Isn't gloo for CPU only training?

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -21,7 +21,7 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 
 ### Fixed
 
-
+- Fixed `dtype` and `device` properties not getting updated in submodules ([#2657](https://github.com/PyTorchLightning/pytorch-lightning/pull/2657))
 
 ## [0.8.5] - 2020-07-09
 

--- a/pytorch_lightning/core/lightning.py
+++ b/pytorch_lightning/core/lightning.py
@@ -68,12 +68,6 @@ class LightningModule(ABC, DeviceDtypeModuleMixin, GradInformation, ModelIO, Mod
         #: True if using amp
         self.use_amp = False
 
-        #: Current dtype
-        self._dtype = torch.float
-
-        #: device reference
-        self._device = torch.device('cpu')
-
         # optionally can be set by user
         self._example_input_array = None
 

--- a/pytorch_lightning/metrics/metric.py
+++ b/pytorch_lightning/metrics/metric.py
@@ -27,8 +27,6 @@ class Metric(DeviceDtypeModuleMixin, torch.nn.Module, ABC):
         """
         super().__init__()
         self.name = name
-        self._dtype = torch.get_default_dtype()
-        self._device = torch.device('cpu')
 
     @abstractmethod
     def forward(self, *args, **kwargs) -> torch.Tensor:

--- a/pytorch_lightning/utilities/device_dtype_mixin.py
+++ b/pytorch_lightning/utilities/device_dtype_mixin.py
@@ -1,4 +1,4 @@
-from typing import Union, Optional
+from typing import Union, Optional, Any, Callable
 
 import torch
 from torch.nn import Module
@@ -82,14 +82,7 @@ class DeviceDtypeModuleMixin(Module):
         """
         # there is diff nb vars in PT 1.5
         out = torch._C._nn._parse_to(*args, **kwargs)
-        device = out[0]
-        dtype = out[1]
-        if device is not None:
-            self.apply(device_apply_fn(device))
-
-        if dtype is not None:
-            self.apply(dtype_apply_fn(dtype))
-
+        self.__update_properties(device=out[0], dtype=out[1])
         return super().to(*args, **kwargs)
 
     def cuda(self, device: Optional[int] = None) -> Module:
@@ -105,7 +98,7 @@ class DeviceDtypeModuleMixin(Module):
         Returns:
             Module: self
         """
-        self.apply(device_apply_fn(torch.device('cuda', index=device)))
+        self.__update_properties(device=torch.device('cuda', index=device))
         return super().cuda(device=device)
 
     def cpu(self) -> Module:
@@ -113,7 +106,7 @@ class DeviceDtypeModuleMixin(Module):
         Returns:
             Module: self
         """
-        self.apply(device_apply_fn(torch.device('cpu')))
+        self.__update_properties(device=torch.device('cpu'))
         return super().cpu()
 
     def type(self, dst_type: Union[str, torch.dtype]) -> Module:
@@ -125,7 +118,7 @@ class DeviceDtypeModuleMixin(Module):
         Returns:
             Module: self
         """
-        self.apply(dtype_apply_fn(dst_type))
+        self.__update_properties(dtype=dst_type)
         return super().type(dst_type=dst_type)
 
     def float(self) -> Module:
@@ -134,7 +127,7 @@ class DeviceDtypeModuleMixin(Module):
         Returns:
             Module: self
         """
-        self.apply(dtype_apply_fn(torch.float))
+        self.__update_properties(dtype=torch.float)
         return super().float()
 
     def double(self) -> Module:
@@ -143,7 +136,7 @@ class DeviceDtypeModuleMixin(Module):
         Returns:
             Module: self
         """
-        self.apply(dtype_apply_fn(torch.double))
+        self.__update_properties(dtype=torch.double)
         return super().double()
 
     def half(self) -> Module:
@@ -152,32 +145,17 @@ class DeviceDtypeModuleMixin(Module):
         Returns:
             Module: self
         """
-        self.apply(dtype_apply_fn(torch.half))
+        self.__update_properties(dtype=torch.half)
         return super().half()
 
+    def __update_properties(self, device=None, dtype=None):
 
+        def apply_fn(module):
+            if not isinstance(module, DeviceDtypeModuleMixin):
+                return
+            if device is not None:
+                module._device = device
+            if dtype is not None:
+                module._dtype = dtype
 
-
-
-def dtype_apply_fn(dtype):
-    return apply_attr('_dtype', dtype)
-
-
-def device_apply_fn(device):
-    return apply_attr('_device', device)
-
-
-# def update_attributes(module, **kwargs):
-#     if not isinstance(module, DeviceDtypeModuleMixin):
-#         return
-#     for k, v in kwargs:
-#         module.__setattr__(k, v)
-
-
-def apply_attr(name: str, value):
-
-    def apply_fn(module):
-        if isinstance(module, DeviceDtypeModuleMixin):
-            module.__setattr__(name, value)
-
-    return apply_fn
+        self.apply(apply_fn)

--- a/pytorch_lightning/utilities/device_dtype_mixin.py
+++ b/pytorch_lightning/utilities/device_dtype_mixin.py
@@ -85,10 +85,10 @@ class DeviceDtypeModuleMixin(Module):
         device = out[0]
         dtype = out[1]
         if device is not None:
-            self._device = device
+            self.apply(device_apply_fn(device))
 
         if dtype is not None:
-            self._dtype = dtype
+            self.apply(dtype_apply_fn(dtype))
 
         return super().to(*args, **kwargs)
 
@@ -105,8 +105,7 @@ class DeviceDtypeModuleMixin(Module):
         Returns:
             Module: self
         """
-
-        self._device = torch.device('cuda', index=device)
+        self.apply(device_apply_fn(torch.device('cuda', index=device)))
         return super().cuda(device=device)
 
     def cpu(self) -> Module:
@@ -114,7 +113,7 @@ class DeviceDtypeModuleMixin(Module):
         Returns:
             Module: self
         """
-        self._device = torch.device('cpu')
+        self.apply(device_apply_fn(torch.device('cpu')))
         return super().cpu()
 
     def type(self, dst_type: Union[str, torch.dtype]) -> Module:
@@ -126,7 +125,7 @@ class DeviceDtypeModuleMixin(Module):
         Returns:
             Module: self
         """
-        self._dtype = dst_type
+        self.apply(dtype_apply_fn(dst_type))
         return super().type(dst_type=dst_type)
 
     def float(self) -> Module:
@@ -135,7 +134,7 @@ class DeviceDtypeModuleMixin(Module):
         Returns:
             Module: self
         """
-        self._dtype = torch.float
+        self.apply(dtype_apply_fn(torch.float))
         return super().float()
 
     def double(self) -> Module:
@@ -144,7 +143,7 @@ class DeviceDtypeModuleMixin(Module):
         Returns:
             Module: self
         """
-        self._dtype = torch.double
+        self.apply(dtype_apply_fn(torch.double))
         return super().double()
 
     def half(self) -> Module:
@@ -153,5 +152,32 @@ class DeviceDtypeModuleMixin(Module):
         Returns:
             Module: self
         """
-        self._dtype = torch.half
+        self.apply(dtype_apply_fn(torch.half))
         return super().half()
+
+
+
+
+
+def dtype_apply_fn(dtype):
+    return apply_attr('_dtype', dtype)
+
+
+def device_apply_fn(device):
+    return apply_attr('_device', device)
+
+
+# def update_attributes(module, **kwargs):
+#     if not isinstance(module, DeviceDtypeModuleMixin):
+#         return
+#     for k, v in kwargs:
+#         module.__setattr__(k, v)
+
+
+def apply_attr(name: str, value):
+
+    def apply_fn(module):
+        if isinstance(module, DeviceDtypeModuleMixin):
+            module.__setattr__(name, value)
+
+    return apply_fn

--- a/pytorch_lightning/utilities/device_dtype_mixin.py
+++ b/pytorch_lightning/utilities/device_dtype_mixin.py
@@ -82,6 +82,10 @@ class DeviceDtypeModuleMixin(Module):
             ExampleModule()
             >>> module.weight #doctest: +ELLIPSIS
             tensor([[...]], dtype=torch.float16)
+            >>> module.device
+            device(type='cpu')
+            >>> module.dtype
+            torch.float16
         """
         # there is diff nb vars in PT 1.5
         out = torch._C._nn._parse_to(*args, **kwargs)

--- a/pytorch_lightning/utilities/device_dtype_mixin.py
+++ b/pytorch_lightning/utilities/device_dtype_mixin.py
@@ -5,8 +5,11 @@ from torch.nn import Module
 
 
 class DeviceDtypeModuleMixin(Module):
-    _device: ...
-    _dtype: Union[str, torch.dtype]
+
+    def __init__(self):
+        super().__init__()
+        self._dtype = torch.get_default_dtype()
+        self._device = torch.device('cpu')
 
     @property
     def dtype(self) -> Union[str, torch.dtype]:

--- a/pytorch_lightning/utilities/device_dtype_mixin.py
+++ b/pytorch_lightning/utilities/device_dtype_mixin.py
@@ -1,4 +1,4 @@
-from typing import Union, Optional, Any, Callable
+from typing import Union, Optional
 
 import torch
 from torch.nn import Module
@@ -148,7 +148,7 @@ class DeviceDtypeModuleMixin(Module):
         self.__update_properties(dtype=torch.half)
         return super().half()
 
-    def __update_properties(self, device=None, dtype=None):
+    def __update_properties(self, device: Optional[torch.device] = None, dtype: Optional[torch.dtype] = None):
 
         def apply_fn(module):
             if not isinstance(module, DeviceDtypeModuleMixin):

--- a/tests/utilities/test_dtype_device_mixin.py
+++ b/tests/utilities/test_dtype_device_mixin.py
@@ -2,6 +2,7 @@ import pytest
 import torch
 import torch.nn as nn
 
+from pytorch_lightning import Trainer, Callback
 from pytorch_lightning.utilities.device_dtype_mixin import DeviceDtypeModuleMixin
 from tests.base import EvalModelTemplate
 
@@ -50,3 +51,24 @@ def test_submodules_device_and_dtype(dst_device, dst_dtype):
     # device and dtype change should propagate down into all children
     assert model.device == model.module.module.device == dst_device
     assert model.dtype == model.module.module.dtype == dst_dtype
+
+
+@pytest.mark.skipif(not torch.cuda.device_count() < 2, reason="test requires multi-GPU machine")
+def test_submodules_multi_gpu_dp(tmpdir):
+
+    class DeviceCallback(Callback):
+
+        def on_batch_start(self, trainer, model):
+            assert isinstance(model, TopModule)
+            assert model.device.index == trainer.local_rank
+            assert model.device == model.module.module.device
+
+    model = TopModule()
+    trainer = Trainer(
+        default_root_dir=tmpdir,
+        distributed_backend='dp',
+        gpus=2,
+        callbacks=[DeviceCallback()],
+        max_steps=1,
+    )
+    trainer.fit(model)

--- a/tests/utilities/test_dtype_device_mixin.py
+++ b/tests/utilities/test_dtype_device_mixin.py
@@ -6,7 +6,7 @@ from pytorch_lightning.utilities.device_dtype_mixin import DeviceDtypeModuleMixi
 from tests.base import EvalModelTemplate
 
 
-class SubSubModule(DeviceDtypeModuleMixin, nn.Module):
+class SubSubModule(DeviceDtypeModuleMixin):
     pass
 
 
@@ -14,14 +14,14 @@ class SubModule(nn.Module):
 
     def __init__(self):
         super().__init__()
-        self.sub_sub_module = SubSubModule()
+        self.module = SubSubModule()
 
 
 class TopModule(EvalModelTemplate):
 
     def __init__(self, *args, **kwargs):
         super().__init__(*args, **kwargs)
-        self.sub_module = SubModule()
+        self.module = SubModule()
 
 
 @pytest.mark.parametrize(['dst_dtype'], [
@@ -44,5 +44,9 @@ def test_submodules_device_and_dtype(dst_device, dst_dtype):
     model = TopModule()
     assert model.device == torch.device('cpu')
     model = model.to(device=dst_device, dtype=dst_dtype)
-    assert model.device == model.sub_module.sub_sub_module.device == dst_device
-    assert model.dtype == model.sub_module.sub_sub_module.dtype == dst_dtype
+    # nn.Module does not have these attributes
+    assert not hasattr(model.module, '_device')
+    assert not hasattr(model.module, '_dtype')
+    # device and dtype change should propagate down into all children
+    assert model.device == model.module.module.device == dst_device
+    assert model.dtype == model.module.module.dtype == dst_dtype

--- a/tests/utilities/test_dtype_device_mixin.py
+++ b/tests/utilities/test_dtype_device_mixin.py
@@ -1,21 +1,48 @@
+import pytest
 import torch
 import torch.nn as nn
 
-from pytorch_lightning.metrics import Accuracy
 from pytorch_lightning.utilities.device_dtype_mixin import DeviceDtypeModuleMixin
 from tests.base import EvalModelTemplate
 
 
-class Model(EvalModelTemplate):
+class SubSubModule(DeviceDtypeModuleMixin, nn.Module):
+    pass
+
+
+class SubModule(nn.Module):
+
+    def __init__(self):
+        super().__init__()
+        self.sub_sub_module = SubSubModule()
+
+
+class TopModule(EvalModelTemplate):
 
     def __init__(self, *args, **kwargs):
         super().__init__(*args, **kwargs)
-        self.acc = Accuracy()
+        self.sub_module = SubModule()
 
 
-def test_submodules_device(tmpdir):
+@pytest.mark.parametrize(['dst_dtype'], [
+    pytest.param(torch.float),
+    pytest.param(torch.double),
+    pytest.param(torch.half),
+])
+@pytest.mark.parametrize(['dst_device'], [
+    pytest.param(torch.device('cpu')),
+    pytest.param(torch.device('cuda')),
+    pytest.param(torch.device('cuda', 0)),
+])
+@pytest.mark.skipif(not torch.cuda.is_available(), reason="test requires GPU machine")
+def test_submodules_device(dst_device, dst_dtype):
+    """
+    Test that the device and dtype property updates propagate through mixed nesting of regular
+    nn.Modules and the special modules of type DeviceDtypeModuleMixin (e.g. Metric or LightningModule).
+    """
 
-    model = Model()
+    model = TopModule()
     assert model.device == torch.device('cpu')
-    model = model.to('cuda')
-    assert model.device == model.acc.device == torch.device('cuda')
+    model = model.to(device=dst_device, dtype=dst_dtype)
+    assert model.device == model.sub_module.sub_sub_module.device == dst_device
+    assert model.dtype == model.sub_module.sub_sub_module.dtype == dst_dtype

--- a/tests/utilities/test_dtype_device_mixin.py
+++ b/tests/utilities/test_dtype_device_mixin.py
@@ -7,7 +7,7 @@ from pytorch_lightning.utilities.device_dtype_mixin import DeviceDtypeModuleMixi
 from tests.base import EvalModelTemplate
 
 
-class SubSubModule(DeviceDtypeModuleMixin):
+class SubSubModule(DeviceDtypeModuleMixin, nn.Module):
     pass
 
 

--- a/tests/utilities/test_dtype_device_mixin.py
+++ b/tests/utilities/test_dtype_device_mixin.py
@@ -56,7 +56,7 @@ def test_submodules_device_and_dtype(dst_device, dst_dtype):
 @pytest.mark.skipif(torch.cuda.device_count() < 2, reason="test requires multi-GPU machine")
 def test_submodules_multi_gpu_dp(tmpdir):
 
-    class DeviceCallback(Callback):
+    class DeviceAssertCallback(Callback):
 
         def on_batch_start(self, trainer, model):
             assert isinstance(model, TopModule)
@@ -68,7 +68,7 @@ def test_submodules_multi_gpu_dp(tmpdir):
         default_root_dir=tmpdir,
         distributed_backend='dp',
         gpus=2,
-        callbacks=[DeviceCallback()],
+        callbacks=[DeviceAssertCallback()],
         max_steps=1,
     )
     trainer.fit(model)

--- a/tests/utilities/test_dtype_device_mixin.py
+++ b/tests/utilities/test_dtype_device_mixin.py
@@ -1,0 +1,21 @@
+import torch
+import torch.nn as nn
+
+from pytorch_lightning.metrics import Accuracy
+from pytorch_lightning.utilities.device_dtype_mixin import DeviceDtypeModuleMixin
+from tests.base import EvalModelTemplate
+
+
+class Model(EvalModelTemplate):
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.acc = Accuracy()
+
+
+def test_submodules_device(tmpdir):
+
+    model = Model()
+    assert model.device == torch.device('cpu')
+    model = model.to('cuda')
+    assert model.device == model.acc.device == torch.device('cuda')

--- a/tests/utilities/test_dtype_device_mixin.py
+++ b/tests/utilities/test_dtype_device_mixin.py
@@ -53,7 +53,7 @@ def test_submodules_device_and_dtype(dst_device, dst_dtype):
     assert model.dtype == model.module.module.dtype == dst_dtype
 
 
-@pytest.mark.skipif(not torch.cuda.device_count() < 2, reason="test requires multi-GPU machine")
+@pytest.mark.skipif(torch.cuda.device_count() < 2, reason="test requires multi-GPU machine")
 def test_submodules_multi_gpu_dp(tmpdir):
 
     class DeviceCallback(Callback):

--- a/tests/utilities/test_dtype_device_mixin.py
+++ b/tests/utilities/test_dtype_device_mixin.py
@@ -7,7 +7,7 @@ from pytorch_lightning.utilities.device_dtype_mixin import DeviceDtypeModuleMixi
 from tests.base import EvalModelTemplate
 
 
-class SubSubModule(DeviceDtypeModuleMixin, nn.Module):
+class SubSubModule(DeviceDtypeModuleMixin):
     pass
 
 

--- a/tests/utilities/test_dtype_device_mixin.py
+++ b/tests/utilities/test_dtype_device_mixin.py
@@ -28,8 +28,10 @@ class TopModule(EvalModelTemplate):
 class DeviceAssertCallback(Callback):
 
     def on_batch_start(self, trainer, model):
+        rank = trainer.local_rank
         assert isinstance(model, TopModule)
-        assert model.device.index == trainer.local_rank
+        # index = None also means first device
+        assert (model.device.index is None and rank == 0) or model.device.index == rank
         assert model.device == model.module.module.device
 
 

--- a/tests/utilities/test_dtype_device_mixin.py
+++ b/tests/utilities/test_dtype_device_mixin.py
@@ -25,6 +25,14 @@ class TopModule(EvalModelTemplate):
         self.module = SubModule()
 
 
+class DeviceAssertCallback(Callback):
+
+    def on_batch_start(self, trainer, model):
+        assert isinstance(model, TopModule)
+        assert model.device.index == trainer.local_rank
+        assert model.device == model.module.module.device
+
+
 @pytest.mark.parametrize(['dst_dtype'], [
     pytest.param(torch.float),
     pytest.param(torch.double),
@@ -55,18 +63,23 @@ def test_submodules_device_and_dtype(dst_device, dst_dtype):
 
 @pytest.mark.skipif(torch.cuda.device_count() < 2, reason="test requires multi-GPU machine")
 def test_submodules_multi_gpu_dp(tmpdir):
-
-    class DeviceAssertCallback(Callback):
-
-        def on_batch_start(self, trainer, model):
-            assert isinstance(model, TopModule)
-            assert model.device.index == trainer.local_rank
-            assert model.device == model.module.module.device
-
     model = TopModule()
     trainer = Trainer(
         default_root_dir=tmpdir,
         distributed_backend='dp',
+        gpus=2,
+        callbacks=[DeviceAssertCallback()],
+        max_steps=1,
+    )
+    trainer.fit(model)
+
+
+@pytest.mark.skipif(torch.cuda.device_count() < 2, reason="test requires multi-GPU machine")
+def test_submodules_multi_gpu_ddp_spawn(tmpdir):
+    model = TopModule()
+    trainer = Trainer(
+        default_root_dir=tmpdir,
+        distributed_backend='dpp_spawn',
         gpus=2,
         callbacks=[DeviceAssertCallback()],
         max_steps=1,

--- a/tests/utilities/test_dtype_device_mixin.py
+++ b/tests/utilities/test_dtype_device_mixin.py
@@ -35,7 +35,7 @@ class TopModule(EvalModelTemplate):
     pytest.param(torch.device('cuda', 0)),
 ])
 @pytest.mark.skipif(not torch.cuda.is_available(), reason="test requires GPU machine")
-def test_submodules_device(dst_device, dst_dtype):
+def test_submodules_device_and_dtype(dst_device, dst_dtype):
     """
     Test that the device and dtype property updates propagate through mixed nesting of regular
     nn.Modules and the special modules of type DeviceDtypeModuleMixin (e.g. Metric or LightningModule).

